Better use of std::string in the debugger symbol_entry class

npwoods opened the pull request.

cuavas commented on src/emu/debug/express.cpp:
```diff
@@ -112,9 +112,9 @@ class integer_symbol_entry : public symbol_entry
-	integer_symbol_entry(symbol_table &table, const char *name, void *ref, symbol_table::getter_func getter, symbol_table::setter_func setter, const std::string &format);
+	integer_symbol_entry(symbol_table &table, std::string &&name, symbol_table::read_write rw, u64 *ptr = nullptr);
+	integer_symbol_entry(symbol_table &table, std::string &&name, u64 constval);
+	integer_symbol_entry(symbol_table &table, std::string &&name, void *ref, symbol_table::getter_func getter, symbol_table::setter_func setter, const std::string &format);
```
If you're trying to cut down on temporaries, why not move the function object as well?

cuavas commented on src/emu/debug/express.cpp:
```diff
@@ -232,26 +232,26 @@ symbol_entry::~symbol_entry()
 
-integer_symbol_entry::integer_symbol_entry(symbol_table &table, const char *name, u64 constval)
-	: symbol_entry(table, SMT_INTEGER, name, "", &m_value),
+integer_symbol_entry::integer_symbol_entry(symbol_table &table, std::string &&name, u64 constval)
```
Constructor delegation - use it.

cuavas commented on src/emu/debug/express.cpp:
```diff
@@ -413,43 +413,54 @@ void symbol_table::configure_memory(void *param, valid_func valid, read_func rea
+void symbol_table::internal_add(std::unique_ptr<symbol_entry> &&symentry)
+{
+	m_symlist.erase(symentry->name());
+	m_symlist.emplace(std::string(symentry->name()), std::move(symentry));
```
Why the extra `std::string`?  It will try to coerce the first parameter to `std::string` anyway when it tries to construct the pair.  Also, why even use this two-step process?  You always want to add/replace an entry if one doesn't exist, and `std::unique_ptr` is both default constructable and move assignable, so you save a hash lookup and potentially an allocation if you do it with a single statement:
```
    m_symlist[symentry->name()] = std::move(symentry);
```

npwoods replied:
Make sense, though the first thing that I ask myself when I see your proposed change is "Is the LHS or the RHS of the assignment evaluated first?  If the RHS is evaluated first, then symentry->name() might fail"

I trust your C++ expertise so I am going to assume that your proposal is safe

cuavas commented on src/emu/debug/express.h:
```diff
@@ -179,10 +179,10 @@ class symbol_table
-	void add(const char *name, void *ref, int minparams, int maxparams, execute_func execute);
+	void add(std::string &&name, read_write rw, u64 *ptr = nullptr);
+	void add(std::string &&name, u64 constvalue);
+	void add(std::string &&name, void *ref, getter_func getter, setter_func setter = nullptr, const std::string &format_string = "");
```
As before, why only move strings?

npwoods commented:
I think that I addressed your feedback, with the caveat that I had to work around what seems to be an obnoxious MSVC bug.

An alternate way to work around this bug that doesn't require MSVC specific #ifdef's is to make the pertinent call to symbol_table::add() be an inline function that takes the functions by value, and have that call a private method that takes the functions as rvalue references.  Let me know if you want to take that approach.

cuavas commented on src/emu/debug/express.cpp:
```diff
@@ -138,7 +141,7 @@ class function_symbol_entry : public symbol_entry
 public:
 	// construction/destruction
-	function_symbol_entry(symbol_table &table, const char *name, void *ref, int minparams, int maxparams, symbol_table::execute_func execute);
+	function_symbol_entry(symbol_table &table, std::string &&name, void *ref, int minparams, int maxparams, symbol_table::execute_func execute);
```
This still takes a temporary function by value and then copies it.

cuavas commented on src/emu/debug/express.cpp:
```diff
@@ -323,8 +326,8 @@ void integer_symbol_entry::internal_setter(symbol_table &table, void *symref, u6
-function_symbol_entry::function_symbol_entry(symbol_table &table, const char *name, void *ref, int minparams, int maxparams, symbol_table::execute_func execute)
-	: symbol_entry(table, SMT_FUNCTION, name, "", ref),
+function_symbol_entry::function_symbol_entry(symbol_table &table, std::string &&name, void *ref, int minparams, int maxparams, symbol_table::execute_func execute)
+	: symbol_entry(table, SMT_FUNCTION, std::move(name), "", ref),
```
Corresponding definition copying the temporary.

cuavas commented on src/emu/debug/express.h:
```diff
@@ -179,10 +179,15 @@ class symbol_table
-	void add(const char *name, void *ref, int minparams, int maxparams, execute_func execute);
+	void add(std::string &&name, read_write rw, u64 *ptr = nullptr);
+	void add(std::string &&name, u64 constvalue);
+#ifdef _MSC_VER
```
What do you mean by "seems not to like" in this case?  It clearly works with rvalue references in many other cases.  Why does it object to this one?

npwoods replied:
I actually don't understand the nature of the bug.  The existence of the overload that takes an std::function as an rvalue reference seems to trigger compilation errors.  I've tried to reproduce the problem in a simpler context, but my attempts have failed.

npwoods commented:
I've hit the other std::function copy, but this seems to also be subject to this apparent MSVC bug.  For reference, here is an example of the compilation error I get when I fail to apply the workaround:

`1>..\..\..\..\..\src\frontend\mame\cheat.cpp(1072): error C2664: 'void symbol_table::add(std::string &&,void *,int,int,symbol_table::execute_func &&)': cannot convert argument 3 from 'int' to 'symbol_table::getter_func'
1>  ..\..\..\..\..\src\frontend\mame\cheat.cpp(1072): note: No constructor could take the source type, or constructor overload resolution was ambiguous
1>..\..\..\..\..\src\frontend\mame\cheat.cpp(1073): error C2664: 'void symbol_table::add(std::string &&,void *,int,int,symbol_table::execute_func &&)': cannot convert argument 3 from 'int' to 'symbol_table::getter_func'
1>  ..\..\..\..\..\src\frontend\mame\cheat.cpp(1073): note: No constructor could take the source type, or constructor overload resolution was ambiguous`

Again, I suppose I could have quietly disregarded the MSVC compilation error and let it bite people, but I feel that would be improper.  Let me know how you want to proceed.
